### DISCUSSION TITLE
How to restore pvc backed up using kopia integration

### DISCUSSION THREAD

**Discussion by adityagu0910** [vmware-tanzu/velero#8165] [Community support Q&A] [answered]
I have created my backup using kopia integration in velero and wanted to know what parameters I have to pass to restore my pvc.

For example, i have deleted my pvc and my deployment and now how do I restore it?

i tried using attached
[restore.txt](https://github.com/user-attachments/files/16805897/restore.txt)
 yaml but it is not restoring the pvc attached to pods/deployments.

**Comment by kaovilai** [comment]
You have specified

```
  includedResources:
  - deployment
```

This mean it will only restore deployment and nothing else.

Removing that from your restore text should restore more items.

**Reply by adityagu0910** [reply]
thanks for the suggestion @kaovilai 
But I want to restore only the deployment/pods and attached pvc with it.
how can I achieve this ?

**Reply by kaovilai** [reply]
First verify what's in backup via backup describe, then if it's only the deployment/pod and attached pvc, you can just restore the entire backup.

Else see https://velero.io/docs/v1.14/resource-filtering/ and include more than just deployment. velero does not auto include other dependent Kinds. If you specify one Kind you require, you must specify the rest you require.

**Comment by adityagu0910** [comment]
@kaovilai 
If I run complete namespace restore then it works fine and my pv, pvc and data inside pvc is getting restored properly. (attached restore yaml)
[onenamespacerestore.txt](https://github.com/user-attachments/files/16909790/onenamespacerestore.txt)

But I wanted to test "only one pvc restore for a namespace which was backed up using kopia integration in velro"
I tried using attached yaml
[restoreonlyonepvcofonenamespace.txt](https://github.com/user-attachments/files/16909808/restoreonlyonepvcofonenamespace.txt)
 also but restore is getting completed but my data is not coming from restore and only k8s resources like pv and pvcs are getting restored.
 
 attached veleronew restore describe restore-nginx-example-pvc-nginx-logs8 --insecure-skip-tls-verify --details o/p 
 
[restoredetailsop.txt](https://github.com/user-attachments/files/16909838/restoredetailsop.txt)

below is what i see in my backup describe : 

`Backup Volumes:
  Velero-Native Snapshots: <none included>

  CSI Snapshots: <none included>

  Pod Volume Backups - kopia:
    Completed:
      nginx-example/nginx-deployment-db9857767-qc5gg: nginx-cache, nginx-logs, nginx-pid`

**Reply by kaovilai** [reply]
> [Velero supports backing up and restoring Kubernetes volumes **attached to pods** from the file system of the volumes, called File System Backup (FSB shortly) or Pod Volume Backup.](https://velero.io/docs/v1.14/file-system-backup/#:~:text=Velero%20supports%20backing%20up%20and%20restoring%20Kubernetes%20volumes%20attached%20to%20pods%20from%20the%20file%20system%20of%20the%20volumes%2C%20called%20File%20System%20Backup%20(FSB%20shortly)%20or%20Pod%20Volume%20Backup.)

Restoring without a pod that references those volumes is guaranteed to break kopia restore.

**Comment by adityagu0910** [comment]
so should I delete deployment, pvc and then include deployment, pvc and pv in restore yaml ?

**Comment by adityagu0910** [comment]
including these three also not restoring the data 

  includedResources:
  - deployments
  - PersistentVolumeClaim
  - PersistentVolume

**Reply by sseago** [reply]
@adityagu0910 you're missing the Pod. The way that fs-backup with kopia works on restore is that the pod is restored with an initcontainer that waits until the filesystem data is restored, then exits and lets the application resume. The PodVolumeRestore resource links to the restored pod and the restored PVC so that the node agent pod running on the same node as the restored pod will have access to the pod's mounted volumes via mount propagation. If the pod isn't restored, then the deployment will redeploy a new pod without this initcontainer, possibly with a completely different name, so there will be no matching pod for the PodVolumeRestore, so the data restore won't work.

If you include pods in that restore, that may be all you need to get this working.

**Comment by adityagu0910** [comment]
@sseago Thanks! that worked.

I have one more query - 
Below steps looks correct to restore only specific deployments(and attached pods) with their pvcs in a namespace which has multiple deployments?
1. Delete deployment and pvcs from namespace (those we want to restore from backup) 
2. Run restore with  pods, deployment, pv and pvc in includedResources like below 

  includedResources:
  - pods
  - deployments
  - PersistentVolumeClaim
  - PersistentVolume
  
3.  This will restore only deleted deployment and pvcs and will not touch all other pvcs and deployments those are already running fine.

**Reply by sseago** [reply, marked as answer]
Yes. If there are other deployments, PVCs, etc in the backup, you will see restore warnings saying that Velero was unable to restore them because they already exist in the cluster, but that's fine since that's what you want here. Note that you should *not* set the `existingResourcePolicy` field on the Restore -- setting it to "update" would tell Velero to attempt to update/modify resources that already exist in the cluster, which could result in kubernetes metadata being changed on your pods/deployments, although it would not cause PV data to be update.

So the short answer is that, yes, as long as you don't set the `existingResourcePolicy` field on the restore, then only deleted (or otherwise not existing) resources from the backup will be restored.

**Comment by adityagu0910** [comment]
okay thanks so attached restore yaml should be good as i am not setting this parameter `existingResourcePolicy`
[restoreonlyonepvcofonenamespace.txt](https://github.com/user-attachments/files/16912016/restoreonlyonepvcofonenamespace.txt)

if we do not specify then default value for this will be `none` ?
